### PR TITLE
impl(generator): add package name support for DiscoveryFile

### LINKED ISSUE #11353: Make generator package aware when emitting protos.
Depending on the package defined in a proto file, some imported types may need to have their names package qualified.

This will involve changes to
- discovery_type_vertex
- discovery_resource
- discovery_to_proto
- discovery_file

### PR DESCRIPTION
**Pull request by scotthart**
fixes #11353

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378)

### CODE REVIEW

File: generator/internal/discovery_file_test.cc [hunk 1]
```diff
@@ -205,32 +205,37 @@ message GetMyResourcesRequest {
...
+  DiscoveryTypeVertex do_foo_request_type("DoFooRequest", "my.package.name",
                                           do_foo_request_type_json);
-  DiscoveryTypeVertex get_request_type("GetMyResourcesRequest", "",
-                                       get_request_type_json);
+  DiscoveryTypeVertex get_request_type(
+      "GetMyResourcesRequest", "my.package.name", get_request_type_json);
   r.AddRequestType("DoFooRequest", &do_foo_request_type);
   r.AddRequestType("GetMyResourcesRequest", &get_request_type);
+  DiscoveryTypeVertex t3("Operation", "other.package", operation_type_json);
+  r.AddResponseType("Operation", &t3);
```
Comment (dbolduc) [inline generator/internal/discovery_file_test.cc:226]: nit: where does the name `t3` come from? Do we want `operation_type` ?

File: generator/internal/discovery_file_test.cc [hunk 2]
```diff
@@ -300,31 +305,36 @@ message GetMyResourcesRequest {
...
```
Comment (dbolduc) [inline generator/internal/discovery_file_test.cc:326]: ditto

File: generator/internal/discovery_file.cc [hunk 3]
```diff
@@ -93,14 +96,19 @@ package $package_name$;
   return {};
 }
 
-Status DiscoveryFile::WriteFile(std::string const& product_name) const {
+Status DiscoveryFile::WriteFile(
+    std::string const& product_name,
+    std::map<std::string, DiscoveryTypeVertex> const* types) const {
+  if (types == nullptr) {
+    return internal::InternalError("types is nullptr");
+  }
```
Comment (dbolduc) [inline generator/internal/discovery_file.cc:104]: comment: It is a little weird to have `types` be a pointer, when `nullptr` is invalid... but I guess you need to put the check somewhere and `WriteFile(...)` is most convenient.

**Review by dbolduc — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @dbolduc)

___
*[`generator/internal/discovery_file.cc` line 104 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378#gh-1176903979:-NTtvM-XC0VPTbKnKclo:b6hmmro) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7def5b5af23328cec5b98553b9ee1bb2863570ec/generator/internal/discovery_file.cc#L104)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

comment: It is a little weird to have `types` be a pointer, when `nullptr` is invalid... but I guess you need to put the check somewhere and `WriteFile(...)` is most convenient.
</blockquote></details>

The reason it's passed as a pointer is because `WriteFile` eventually gets called in a `std::async` and the `types` pointer gets passed as one of the arguments to `std::async`.  Perhaps its overly cautious, but I didn't want to pass a reference.
___
*[`generator/internal/discovery_file_test.cc` line 226 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378#gh-1176895147:-NTtxvUUCxVz_odkUJnX:bc1ztvw) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7def5b5af23328cec5b98553b9ee1bb2863570ec/generator/internal/discovery_file_test.cc#L226)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: where does the name `t3` come from? Do we want `operation_type` ?
</blockquote></details>

When I copy/pasted it from another test file it was t3 as its predecessors where t1 and t2. Updated the name for this file. Changing the name in this file to operation_type.
___
*[`generator/internal/discovery_file_test.cc` line 326 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378#gh-1176896307:-NTty4Zb4KpnEeFb7qhz:b499q7j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7def5b5af23328cec5b98553b9ee1bb2863570ec/generator/internal/discovery_file_test.cc#L326)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Changed
___
*[`generator/internal/discovery_file_test.cc` line 557 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378#gh-1176896519:-NTty5W-5YnYitdpWLbT:b499q7j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7def5b5af23328cec5b98553b9ee1bb2863570ec/generator/internal/discovery_file_test.cc#L557)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Changed

**Comment by coryan** [comment]
*[`generator/internal/discovery_file.cc` line 104 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378#gh-1176903979:-NTtz0bH96IM7dgkKFvp:b-m1a0y5) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7def5b5af23328cec5b98553b9ee1bb2863570ec/generator/internal/discovery_file.cc#L104)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

The reason it's passed as a pointer is because `WriteFile` eventually gets called in a `std::async` and the `types` pointer gets passed as one of the arguments to `std::async`.  Perhaps its overly cautious, but I didn't want to pass a reference.
</blockquote></details>

Optional: consider `std::ref()` or `std::cref()`.  They are designed precisely to pass references to `std::async()` and similar functions.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[`generator/internal/discovery_file.cc` line 104 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11378#gh-1176903979:-NTu-o5n-HfYuIy2tbgR:b-j8sb1u) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7def5b5af23328cec5b98553b9ee1bb2863570ec/generator/internal/discovery_file.cc#L104)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Optional: consider `std::ref()` or `std::cref()`.  They are designed precisely to pass references to `std::async()` and similar functions.
</blockquote></details>

I'll update it in the next PR.

**Pull request merged**